### ISSUE TITLE
Evaluated expressions of variables differ sometimes when using the GPU

### ISSUE THREAD

**Issue by alexlee-gk** [tensorflow/tensorflow#2226]
The evaluated value of the l1 or l2 loss of variables differ sometimes when using the GPU. This seems to happen when the variables are "large".

Even though the difference is not that much in the example below, these differences have resulted in a 98% test accuracy on a network trained on the CPU, but a 10% test accuracy on the same network trained on the GPU.

### Environment info
Operating System: Ubuntu 14.04

Installed version of CUDA and cuDNN: 7.5 and 7.0 [cuda_info.txt](https://github.com/tensorflow/tensorflow/files/249768/cuda_info.txt)
GPU: Titan X

If installed from binary pip package, provide:
1. Which pip package you installed.
pip install --upgrade https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.8.0-cp27-none-linux_x86_64.whl
2. The output from python -c "import tensorflow; print(tensorflow.__version__)".
python -c "import tensorflow; print(tensorflow.__version__)"I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcuda.so.1 locally
I tensorflow/stream_executor/dso_loader.cc:105] successfully opened CUDA library libcurand.so locally
0.8.0

### Steps to reproduce
    import tensorflow as tf
    W1 = tf.Variable(tf.truncated_normal([5, 5, 1, 32], stddev=0.1))
    W2 = tf.Variable(tf.truncated_normal([5, 5, 1, 32], stddev=0.1))
    sess = tf.InteractiveSession()
    sess.run(tf.initialize_all_variables())
    l2 = tf.reduce_sum(W1 * W1) + tf.reduce_sum(W2 * W2)
    for i in range(100):
        print('%.10f' % sess.run(l2))

**Comment by zheng-xq** [comment]
The GPU kernel for reduce_sum is known to be non-deterministic since it
uses the atomic operations. When the number of elements is large, the
difference could be quite large.

If your model is trained with dropout in it, it tends to be less likely to
be affected by the noise.

Otherwise, it is a good idea to have a small example where we can observe
the 98% -> 10% difference.

**Comment by alexlee-gk** [comment]
I have condensed my code into a single script that builds a small convnet and trains it for 1000 iterations. The network achieves a test accuracy of 97.7% and 9.8% when I use the CPU and GPU respectively.

I have also attached weights for initializing the network; this particular initialization seems to be particularly bad even though the non-bias weights were initialized using a truncated normal distribution.

Another thing to note is that this behavior happens with a particular bad choice of hyperparameters (learning rate 0.1 and momentum 0.1).

I'm aware that a lot of things could be better in the script, such as using a lower learning rate, higher momentum, dropout, batch norm, etc. However, I'm interested in this particular setting since this is what we developed for a class that I'm teaching (CS188 at UC Berkeley), and the behavior of our online autograder differs from that of the student's local autograder. Another puzzling thing is that the online autograder is the one getting ~10% accuracy even though it is not using the GPU. This autograder is running Ubuntu 14.04 on aws.

[convnet_determinism_check.zip](https://github.com/tensorflow/tensorflow/files/249904/convnet_determinism_check.zip)

**Comment by alexlee-gk** [comment]
Update: If the network is not initialized with the provided weights, but instead initialized using the truncated normal distribution (for the non-bias parameters), the network achieves around 95% test accuracy, which still seems significant.

**Comment by girving** [comment]
@zheng-xq: Any thoughts?  @alexlee-gk: If not, we may have to close this as expected behavior, unfortunately.  It sounds like your 97.7% vs. 9.8% difference occurs for a model where different CPU systems can also produce 10%, and thus that you may have found an impressively unreliable set of hyperparameters.  If using a smaller learning rate helps, it may just mean that the GPU version blows up slightly earlier than the CPU version.

**Comment by zheng-xq** [comment]
@alexlee-gk, the problem could be attributed to small differences introduced in the non-deterministic behavior on GPU. Or some numerical instability due to the precision difference between CPU and GPU. 

The best way to tell whether this is actually a bug is to narrow them into a small script that shows a big difference in one or two steps. If that is not possible, then the model might be unstable, and this might be an expected behavior.

**Comment by alexlee-gk** [comment]
I think it is reasonable to attribute the problem as an expected behavior, which is caused due to numerical instability.

That being said, I just want to point out that this behavior is not limited to GPU builds. There were sets of hyperparameters in which I would get high accuracy in Mac OSX (CPU) and Ubuntu 14.04 (CPU), but a lower accuracy in Ubuntu 14.04 running on AWS (CPU).

**Comment by zheng-xq** [comment]
@alexlee-gk, thanks for bringing this issue to our attention. 

Closing as expected behavior. Feel free to reopen if you can reproduce with large difference within one or two steps.

**Comment by yaroslavvb** [comment]
At some point I was debugging tests failing when running with -copt=avx2 instead of avx. We couldn't get rid of the numeric difference and ended up just raising tolerances on tests. What happens is that there's some atomic operation that produces small difference in the range of ulp or two and then it blows up. In particular, I noticed forward pass during MNIST training to be relatively stable, but during backward pass, tiny difference blow up to billions of ulps
